src/util.rs

```diff
@@ -15,3 +18,163 @@ pub fn ptrdistance<T>(a: *const T, b: *const T) -> usize {
     debug_assert!(a as usize <= b as usize);
     (b as usize - a as usize) / size_of::<T>()
 }
+
+// Trait for calling &slice[r] for range r on either shared or mutable slices
+pub trait SliceWithRange {
+    fn slice_range(self, r: Range<usize>) -> Self;
+    fn slice_rangefrom(self, r: RangeFrom<usize>) -> Self;
+    fn slice_rangeincl(self, r: RangeInclusive<usize>) -> Self;
+}
+
+impl<'a, T> SliceWithRange for &'a [T] {
+    fn slice_range(self, r: Range<usize>) -> Self { &self[r] }
+    fn slice_rangefrom(self, r: RangeFrom<usize>) -> Self { &self[r] }
+    fn slice_rangeincl(self, r: RangeInclusive<usize>) -> Self { &self[r] }
+}
+
+impl<'a, T> SliceWithRange for &'a mut [T] {
+    fn slice_range(self, r: Range<usize>) -> Self { &mut self[r] }
+    fn slice_rangefrom(self, r: RangeFrom<usize>) -> Self { &mut self[r] }
+    fn slice_rangeincl(self, r: RangeInclusive<usize>) -> Self { &mut self[r] }
+}
+
+/// Fluent API for slicing with RangeBounds on vectors and slices
+pub trait SliceWith {
+    fn slice<I>(&self, r: I) -> &Self where I: RangeBounds<usize>;
+    fn slice_mut<I>(&mut self, r: I) -> &mut Self where I: RangeBounds<usize>;
+}
+
+impl<T> SliceWith for [T] {
+    fn slice<I>(&self, r: I) -> &Self where I: RangeBounds<usize>
+    { slice_by(self, r) }
+
+    fn slice_mut<I>(&mut self, r: I) -> &mut Self where I: RangeBounds<usize>
+    { slice_by(self, r) }
+}
+
+fn slice_by<S, I>(v: S, range: I) -> S
+    where S: SliceWithRange, I: RangeBounds<usize>
+{
+    let start = match range.start_bound() {
+        Bound::Unbounded => 0,
+        Bound::Included(&i) => i,
+        Bound::Excluded(&i) => check_exclusive_to_inclusive_start(i),
+    };
+    match range.end_bound() {
+        Bound::Excluded(&i) => v.slice_range(start..i),
+        Bound::Included(&i) => v.slice_rangeincl(start..=i),
+        Bound::Unbounded => v.slice_rangefrom(start..),
+    }
+}
+
+/// Compute i + 1 but check for overflow
+#[inline]
+fn check_exclusive_to_inclusive_start(i: usize) -> usize {
+    // Range exclusive from usize max is always out of bounds.
+    // Exclusive from usize max to unbounded is a well-formed range (!0, infinity)
+    // but always out of bounds in slices which have len <= !0 anyway.
+    //
+    // Range exclusive from usize max to Included/Excluded usize value is
+    // an ill-formed range because it does not have start <= end, but we
+    // don't make the difference here.
+    //
+    // start at i + 1 but panic for oob on overflow
+    let (iplus1, overflow) = i.overflowing_add(1);
+    if overflow {
+        panic!(concat!("Range out of bounds: exclusive from ",
+                stringify!(std::usize::MAX)));
```

Comment:
That `stringify!` prints the tokens, not the constant value -- `std :: usize :: MAX`. You could use a runtime formatting string for `panic!` -- this is a cold case anyway. Or just say, "exclusive from usize::MAX".

Reply:
printing the tokens was intended, but the whitespace isn't pretty. I guess I can just use the simple option of writing usize::MAX :)